vernac/declare.mli

```diff
@@ -198,7 +198,10 @@ module Proof : sig
     -> int list option
     -> t
 
-  (** Qed a proof  *)
+  (** Qed or Define a proof. This API expects to run in an environment where
+      side effects generated by the tactics which build the proof are present,
+      that is the state has not been rewound to the proof start (as the STM
+      does).  *)
```

Comment:
Ok, so both for `Defined` and `Qed` the input env has to be "linear" one.

But we still have cases in the CI for the simple compiler where this is not working properly, right?

For example, https://github.com/coq/coq/pull/16361/checks?check_run_id=7774483046 is with this PR applied right?

Reply:
This PR repairs some cases which are tested here and in the simple compiler about side effects being displayed as axioms. Some of these cases are also visible without the simple compiler, hence this PR is closing an existing issue.

This PR does not solve all problems we have with the simple compiler, nor claims of doing so.

In passing is simplifies the kernel.

Reply:
Ok, so indeed we don't yet met the contract, but at least we now have a intention about what the contract should be :+1: